table/tables/partition.go

```diff
@@ -0,0 +1,163 @@
+}
+
+// PartitionExpr is the partition definition expressions.
+// There are two expressions exist, because Locate use binary search, which requires:
+// Given a compare function, for any partition range i, if cmp[i] > 0, then cmp[i+1] > 0.
+// While partition prune must use the accurate range to do prunning.
+// partition by range (x)
+//   (partition
+//      p1 values less than (y1)
+//      p2 values less than (y2)
+//      p3 values less than (y3))
+// PartitionPrune: (x < y1); (y1 <= x < y2); (y2 <= x < y3)
+// Locate: (x < y1); (x < y2); (x < y3)
+type PartitionExpr struct {
+	PartitionPrune []expression.Expression
+	Locate         []expression.Expression
+}
+
+func generatePartitionExpr(tblInfo *model.TableInfo) (*PartitionExpr, error) {
+	// The caller should assure partition info is not nil.
+	pi := tblInfo.GetPartitionInfo()
+	ctx := mock.NewContext()
+	partitionPruneExprs := make([]expression.Expression, 0, len(pi.Definitions))
+	locateExprs := make([]expression.Expression, 0, len(pi.Definitions))
+	var buf bytes.Buffer
+	for i := 0; i < len(pi.Definitions); i++ {
+		if strings.EqualFold(pi.Definitions[i].LessThan[0], "MAXVALUE") {
+			// Expr less than maxvalue is always true.
+			fmt.Fprintf(&buf, "true")
+		} else {
+			fmt.Fprintf(&buf, "((%s) < (%s))", pi.Expr, pi.Definitions[i].LessThan[0])
+		}
+		expr, err := expression.ParseSimpleExpr(ctx, buf.String(), tblInfo)
+		if err != nil {
+			// If it got an error here, ddl may hang forever, so this error log is important.
+			log.Error("wrong table partition expression:", errors.ErrorStack(err), buf.String())
+			return nil, errors.Trace(err)
+		}
+		locateExprs = append(locateExprs, expr)
+
+		if i > 0 {
+			fmt.Fprintf(&buf, " and ((%s) >= (%s))", pi.Expr, pi.Definitions[i-1].LessThan[0])
+		}
+
+		expr, err = expression.ParseSimpleExpr(ctx, buf.String(), tblInfo)
+		if err != nil {
+			// If it got an error here, ddl may hang forever, so this error log is important.
+			log.Error("wrong table partition expression:", errors.ErrorStack(err), buf.String())
+			return nil, errors.Trace(err)
+		}
+		partitionPruneExprs = append(partitionPruneExprs, expr)
+		buf.Reset()
+	}
+	return &PartitionExpr{
+		PartitionPrune: partitionPruneExprs,
+		Locate:         locateExprs,
+	}, nil
+}
+
+// PartitionExpr returns the partition expression.
+func (t *PartitionTable) PartitionExpr() *PartitionExpr {
+	return t.partitionExpr
+}
+
+func partitionRecordKey(pid int64, handle int64) kv.Key {
+	recordPrefix := tablecodec.GenTableRecordPrefix(pid)
+	return tablecodec.EncodeRecordKey(recordPrefix, handle)
+}
+
+// locatePartition returns the partition ID of the input record.
+func (t *PartitionTable) locatePartition(ctx sessionctx.Context, pi *model.PartitionInfo, r []types.Datum) (int64, error) {
+	var err error
+	partitionExprs := t.partitionExpr.Locate
+	idx := sort.Search(len(partitionExprs), func(i int) bool {
+		var ret int64
+		ret, _, err = partitionExprs[i].EvalInt(ctx, types.DatumRow(r))
```

Comment:
This is a binary search? And you want to find the first i which could satisfy cmp[i] > 0?

Reply:
Bingo!

Reply:
There may be more than one `i` could satisfy the condition `cmp[i] > 0`.  Could the binary search guarantee find the first one?

Reply:
Yes. From the `sort.Search` doc:
https://godoc.org/sort#Search

> Search uses binary search to find and return the smallest index i in [0, n) at which f(i) is true, **assuming that on the range [0, n), f(i) == true implies f(i+1) == true**. That is, Search requires that f is false for some (possibly empty) prefix of the input range [0, n) and then true for the (possibly empty) remainder; **Search returns the first true index**.

Reply:
Got it.

Reply:
I see.